ext/lib/encoding/tinycbor/src/compilersupport_p.h

```diff
@@ -0,0 +1,263 @@
+/****************************************************************************
+**
+** Copyright (C) 2017 Intel Corporation
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy
+** of this software and associated documentation files (the "Software"), to deal
+** in the Software without restriction, including without limitation the rights
+** to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+** copies of the Software, and to permit persons to whom the Software is
+** furnished to do so, subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in
+** all copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+** THE SOFTWARE.
+**
+****************************************************************************/
+
+#ifndef COMPILERSUPPORT_H
+#define COMPILERSUPPORT_H
+
+#include "cbor.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef _BSD_SOURCE
+#  define _BSD_SOURCE
+#endif
+#ifndef _DEFAULT_SOURCE
+#  define _DEFAULT_SOURCE
+#endif
+#ifndef assert
+#  include <assert.h>
+#endif
+#include <float.h>
+#include <math.h>
```

Comment:
zephyr does not have this in its minimal libc, so you will have to make tinycbor depend on newlib libc, otherwise build will fail. If we had a simple test using the library, this would have been identified as an issue earlier.

Reply:
Agreed, mcumgr also depends on newlib libc. CC @nvlsianpu @ccollins476ad

Reply:
float.h should be available everywhere. It's required by the C standard even in freestanding implementations and it is provided by gcc, not the libc.

math.h is used in the `decode_half` function and that one only uses `ldexp`. I will accept a Pull Request that reimplements the function without `ldexp`, like `encode_half` above. If anyone wants to contribute an ARM `__fp16` implementation like the x86 F16C, I'll also accept it.

Reply:
assert.h also seems to be in the toolchain:

```
$ find /opt/zephyr-sdk/ -name assert.h
/opt/zephyr-sdk/sysroots/mips32r2-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/arc-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/riscv32-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/i586-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/nios2-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/iamcu-zephyr-elfiamcu/usr/include/assert.h
/opt/zephyr-sdk/sysroots/xtensa-zephyr-elf/usr/include/assert.h
/opt/zephyr-sdk/sysroots/armv5-zephyr-eabi/usr/include/assert.h
```

It is compliant with C11. It does call a function called `__assert_func` on failed assertions, but if you don't want hat you can always compile with NDEBUG.

Reply:
@ccollins476ad @vrahane you mentioned you are building this without newlib?

Reply:
@carlescufi mcumgr doesn't require any of the TinyCBOR C files that include `math.h` (`cbortojson.c`, `cborpretty.c`, and `cborvalidation.c`)[*].  Removing these files from the TinyCBOR `CMakeLists.txt` file allowed a successful build without pulling in newlib.

Since TinyCBOR is being added as an external zephyr library, and not just an external mcumgr library, we can't just remove these files from the `CMakeLists.txt` file.  Perhaps a new Kconfig setting to exclude these files is required.

[*] `compilersupport_p.h` also includes `math.h`.  I'm not sure why this didn't cause build failures, as several source files include this header.  I don't have a suitable machine available at the moment, but I will investigate this later today and include my findings here.

Reply:
I agree with @ccollins476ad above. We might have to investigate this further but adding a config option to depend on does seem like the right approach.

Reply:
I looked at this again, and it seems my compiler[1] has no problem with `#include <math.h>`, even when `CONFIG_NEWLIB_LIBC` is disabled.

I am pretty sure I used to get a build error due to the missing `math.h` header, but now I can't reproduce it.  Maybe I just remembered incorrectly...

[1] `gcc version 5.4.1 20160919 (release) [ARM/embedded-5-branch revision 240496] (GNU Tools for ARM Embedded Processors)`

Reply:
@ccollins476ad can you please try with the Zephyr SDK: http://docs.zephyrproject.org/getting_started/installation_linux.html#installing-the-zephyr-software-development-kit

Reply:
Unfortunately, I do get build errors with the Zephyr SDK:
```
In file included from /home/ccollins/repos/zephyr/ext/lib/encoding/tinycbor/src/cbor_buf_reader.c:26:0:
/home/ccollins/repos/zephyr/ext/lib/encoding/tinycbor/src/compilersupport_p.h:44:18: fatal error: math.h: No such file or directory                                    
 #include <math.h>                                                                                                                                                     
                  ^                                                                                                                                                    
compilation terminated.                                                                                                                                                
```

Reply:
I don't see math.h anywhere in the Zephyr sources or SDK. Like I said, the only function used is ldexp and that was only so I wouldn't depend on IEEE 754 internals (like the other function does).

In any case, as  @ccollins476ad said, the `decode_half` function is only used in the cborpretty.c, cborvalidation.c, and cbortojson.c, which aren't used, so you can just erase the function and the include, for now.

(That would make it my problem to fix when integrating with Zephyr...)